base/inference.jl

```diff
@@ -2120,13 +2121,139 @@ function inlineable(f, e::Expr, atypes, sv, enclosing_ast)
     if isa(f,IntrinsicFunction)
         return NF
     end
+    if is(f, invoke)
+        return inlineable_invoke(f, e, atypes, sv, enclosing_ast, argexprs)
+    end
+
+    return inlineable_gf(f, e, atypes, sv, enclosing_ast, argexprs)
+end
+
+function get_invoke_types{T<:Top}(ts::Type{Type{T}})
+    return T
+end
+
+function get_invoke_types(ts::Tuple)
+    return tuple(Any[get_invoke_types(t) for t in ts]...)::Type
+end
+
+function inlineable_invoke(f::Function, e::Expr, atypes, sv,
+                           enclosing_ast, argexprs)
+    f = isconstantfunc(argexprs[1], sv)
+    if is(f, false)
+        return NF
+    end
+    f = _ieval(f)
+    local invoke_types
+    try
+        invoke_types = get_invoke_types(atypes[2])
+    catch err
+        return NF
+    end
 
+    argexprs = argexprs[3:end]
+    atypes = atypes[3:end]
+    if length(atypes) != length(invoke_types)
+        return NF
+    end
+    if isleaftype(invoke_types) && invoke_types == atypes
+        new_e = Expr(:call, f, argexprs...)
+        new_e.typ = e.typ
+        res = inlineable_gf(f, new_e, invoke_types, sv,
+                            enclosing_ast, argexprs)
+        return is(res, NF) ? (new_e, []) : res
+    end
+
+    local meth
+    try
+        meth = ccall(:jl_gf_invoke_lookup, Any, (Any, Any), f, invoke_types)
+    catch
+        return NF
+    end
+    if is(meth, nothing)
+        return NF
+    end
+
+    stmts = []
+    check_stmts = []
+    atypes_l = Type[atypes...]
+    err_label = genlabel(sv)
+    after_err_label = genlabel(sv)
+    for i in 1:length(atypes_l)
```

Comment:
The type check is done after all arguments are evaluated so changing to `(Any, Integer)` does not affect the evaluation of the arguments at all. (which is the same schematics with calling `invoke` function)

(Actually I think I might miss the case where evaluating the second argument to `invoke` has side effect)....